api/src/main/java/io/opentelemetry/trace/Tracestate.java

```diff
@@ -235,28 +235,29 @@ public static Entry create(String key, String value) {
 
   // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
   // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
-  // forward slashes /.
+  // forward slashes /.  For multi-tenant vendor scenarios, an at sign (@) can be used to prefix the
+  // vendor name.
   private static boolean validateKey(String key) {
-    if (key.length() > KEY_MAX_SIZE
-        || key.isEmpty()
-        || key.charAt(0) < 'a'
-        || key.charAt(0) > 'z') {
+    if (key.length() > KEY_MAX_SIZE || key.isEmpty() || isNotNumberOrDigit(key.charAt(0))) {
       return false;
     }
+    int atSeenCount = 0;
     for (int i = 1; i < key.length(); i++) {
       char c = key.charAt(i);
-      if (!(c >= 'a' && c <= 'z')
-          && !(c >= '0' && c <= '9')
-          && c != '_'
-          && c != '-'
-          && c != '*'
-          && c != '/') {
+      if (isNotNumberOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/') {
```

Comment:
Add a TODO that it may worth having an array `bool[256] validBytes` that will remove a lot of jumps here.

Reply:
I would push back on this level of detailed optimization this early in a pre-alpha release cycle.  TODOs like that always clutter up the code, and a distant future developer optimizing this area of the code might have an even better approach.

In any case, the current code is walking `char`s not `byte`s so we'd want `boolean[65536]` to index with char or assume an 8-bit encoding and convert char to byte and hope that we don't get a negative byte...

I dunno, doesn't seem worth it to me, especially this early.

Reply:
I would very much counsel against littering the code with that type of comment. Remember - almost by definition - the code which matters is that which comes out of the JIT. It will not look like this.

Reply:
Thanks to @adriancole for raising this issue and help.

As much as I am a believer in the JIT magic (not) I would like to see JMH benchmarks to prove me wrong. As I suggested I did not say to do it, just as an optimization opportunity and a way to do it.

I am not going to argue that it is good or bad suggestion because that is something we can prove with benchmarks (and yes I believe in benchmarks).

I don't think in general a to-do litter the code, a well written todo may explain why an implementation is done in a specific way or the other.

So I would suggest that we either benchmark now between the two implementation or we add a comment that we did not benchmark this but an alternative implementation may be considered ....

Reply:
I added a todo to benchmark the key validation code.